MAE_WORKDIR = cfg.MAE.getWorkdir(str_=False)

MAE_index_name = "mae-pipeline"
MAE_index_input, MAE_index_output, MAE_graph_file, _ = createIndexRule(
    scriptsPath=str(MAE_WORKDIR),
    index_name=MAE_index_name
)

def mae_Overview_R_input(cfg):
    if cfg.MAE.run:
	 10 #'    - allelic_counts: '`sm expand(cfg.getProcessedDataDir() +
 11 #'                          "/mae/allelic_counts/{mae_id}.csv.gz",
 12 #'                          mae_id=cfg.MAE.getMaeAll())`'
 13 #'    - results_obj: '`sm expand(cfg.getProcessedResultsDir() +
 14 #'                       "/mae/samples/{mae_id}_res.Rds",
 15 #'                       mae_id=cfg.MAE.getMaeAll())`'
 16 #'    - results_tables: '`sm expand(cfg.getProcessedResultsDir() +
"/mae/{dataset}/MAE_results_{annotation}.tsv",
dataset=cfg.MAE.groups, annotation=cfg.genome.getGeneVersions())`'

qc_matrix: '`sm expand(cfg.getProcessedResultsDir() + "/mae/{qc_group}/" +
"dna_rna_qc_matrix.Rds", qc_group=cfg.MAE.qcGroups)`'

       return {"allelic_counts" :
                                  expand(cfg.getProcessedResultsDir() + "/mae/allelic_counts/{mae_id}.csv.gz",
                                  mae_id=cfg.MAE.getMaeAll()),
			  "results_obj": 
			                      expand(cfg.getProcessedResultsDir() + "/mae/samples/{mae_id}_res.Rds,
                                  mae_id=cfg.MAE.getMaeAll()),

              "result_tables":
                                  expand(cfg.getProcessedResultsDir() + "/mae/{dataset}/MAE_results_{annotation}.tsv",
                                  "/aberrant_splicing/results/{annotation}/fraser/{dataset}/results_per_junction.tsv",
                                  dataset=cfg.MAE.groups, annotation=cfg.genome.getGeneVersions()),
              "qc_matrix":        
			                      expand(cfg.getProcessedResultsDir() + "/mae/{qc_group}/" + "dna_rna_qc_matrix.Rds",
								  qc_group=cfg.MAE.qcGroups)
                }
    else:
        return {"dummy_file": cfg.getProcessedDataDir() + "/.module_skips/mae.skip"}




rule mae:
    input: MAE_index_input, MAE_graph_file
    output: MAE_index_output
    run: ci(str(MAE_WORKDIR), MAE_index_name)

rule mae_dependency:
    output: MAE_graph_file
    shell: "snakemake --nolock --rulegraph {MAE_index_output} | dot -Tsvg -Grankdir=TB > {output}"

rule sampleQC:
    input: cfg.getHtmlFromScript(MAE_WORKDIR / "QC" / "Datasets.R")

rule create_dict:
    input:  "{fasta_file_base}.fa"
    output: "{fasta_file_base}.dict"
    params:
        gatk = cfg.getTool("gatkCmd"),
    shell: "{params.gatk} CreateSequenceDictionary --REFERENCE {input}"

## MAE
rule mae_createSNVs:
    input:
        ncbi2ucsc = MAE_WORKDIR / "resource" / "chr_NCBI_UCSC.txt",
        ucsc2ncbi = MAE_WORKDIR / "resource" / "chr_UCSC_NCBI.txt",
        vcf_file  = lambda w: cfg.MAE.getVcf(w.vcf),
        bam_file  = lambda w: sa.getFilePath(w.rna, 'RNA_BAM_FILE'),
        script    = MAE_WORKDIR / "MAE" / "filterSNVs.sh",
        snv_params = lambda w: cfg.processedDataDir / "mae" / "params/snvs" / (w.rna + "_snvParams.csv")
    output:
        snvs_filename = cfg.processedDataDir / "mae" / "snvs" / "{vcf}--{rna}.vcf.gz",
        snvs_index = cfg.processedDataDir / "mae" / "snvs" / "{vcf}--{rna}.vcf.gz.tbi"
    params:
        bcftools = cfg.getTool("bcftoolsCmd"),
        samtools = cfg.getTool("samtoolsCmd")
    shell:
        """
        {input.script} {input.ncbi2ucsc} {input.ucsc2ncbi} {input.vcf_file} \
        {wildcards.vcf} {input.bam_file} {output.snvs_filename} \
        {params.bcftools} {params.samtools}
        """

rule mae_allelicCounts:
    input:
        ncbi2ucsc = MAE_WORKDIR / "resource" / "chr_NCBI_UCSC.txt",
        ucsc2ncbi = MAE_WORKDIR / "resource" / "chr_UCSC_NCBI.txt",
        vcf_file  = rules.mae_createSNVs.output.snvs_filename,
        bam_file  = lambda w: sa.getFilePath(w.rna, 'RNA_BAM_FILE'),
        fasta     = lambda w: cfg.MAE.getGenomePath(w.rna),
        dict      = lambda w: cfg.genome.getFastaDict(cfg.MAE.getGenomePath(w.rna)),
        script    = MAE_WORKDIR / "MAE" / "ASEReadCounter.sh"
    output:    
        counted = cfg.processedDataDir / "mae" / "allelic_counts" / "{vcf}--{rna}.csv.gz"
    params:
        bcftools = cfg.getTool("bcftoolsCmd"),
        samtools = cfg.getTool("samtoolsCmd"),
        gatk = cfg.getTool("gatkCmd"),
        gatkIgnoreHeaderCheck = cfg.MAE.get("gatkIgnoreHeaderCheck")
    shell:
        """
        {input.script} {input.ncbi2ucsc} {input.ucsc2ncbi} \
        {input.vcf_file} {input.bam_file} {wildcards.vcf}--{wildcards.rna} \
        {input.fasta} {params.gatkIgnoreHeaderCheck} {output.counted} \
        {params.bcftools} {params.samtools} {params.gatk}
        """
